🚧 Split luminol into separate crates

melody-rs opened the pull request:
This is currently WIP. The goal of this PR is to split Luminol into separate crates and ditch interior mutability, hopefully solving a whole class of bugs related to opening and closing projects.

melody-rs committed:
refactor: :construction: Split luminol into separate crates

melody-rs committed:
refactor: :construction: Start working out dependencies

melody-rs committed:
Sorta figure out dependencies

melody-rs committed:
Sorta start getting somewhere

melody-rs committed:
Move Luminol into crates folder

melody-rs committed:
Move the filesystem trait out of luminol-core

melody-rs committed:
refactor: :construction: refactor luminol-graphics and it's a mess

melody-rs committed:
refactor:

melody-rs committed:
refactor: add UpdateState to luminol-core and move various traits

melody-rs committed:
refactor(filesystem): :recycle: require FileSystem::File to be 'static

melody-rs committed:
refactor: :recycle: don't take an &'static reference to graphics state

melody-rs committed:
refactor: :recycle: refactor modals

melody-rs committed:
refactor: :recycle: unify tabs and windows

melody-rs committed:
refactor: :recycle: start to fix up windows

melody-rs committed:
refactor: :recycle: hack together things so that they compile

melody-rs committed:
Merge branch 'dev' into split-crates

melody-rs committed:
refactor: :recycle: partially resolve async code issues

melody-rs committed:
refactor: :recycle: it compiles (with a LOT of unfinished things)

melody-rs committed:
refactor: :recycle: NOW it compiles

melody-rs committed:
fix(graphics): :bug: fix sprite shader compilation

melody-rs committed:
fix(data cache): :recycle: get data cache semi-working

melody-rs committed:
refactor(data cache): :recycle: actually load data cache

melody-rs committed:
refactor(config): :recycle: store code theme in global state again

melody-rs committed:
fix(ui): :bug: fix windows and tabs not being added

melody-rs committed:
refactor: :recycle: use workspace metadata for package metadata

melody-rs committed:
refactor: :recycle: remove generic parameters on update state by using dynamic dispatch

I couldn't get generics to allow adding tabs or windows inside tabs or windows

melody-rs committed:
refactor(tabs): :recycle: pass update state when requesting tab name

melody-rs committed:
fix(tabs): :bug: fix tabs not adding properly

melody-rs committed:
Update window.rs

melody-rs committed:
fix: :bug: get top bar loading projects

melody-rs committed:
fix: :bug: fix new project creation

because reqwest requires tokio we need to spawn a tokio runtime. i do not think this scales well to web

melody-rs commented:
While this PR is not complete it is almost finished, so I'm opening it up for review.

Things I need to still do:
- Improve error contexts and error handling
- Convert the data cache into an option
- Perform more testing
- Fix excessive function parameters (builder pattern?)
- Fix webassembly support

This PR removes some otherwise unfinished features (namely the event editor and item editor)
They are out of the scope of this pr and will be reimplemented in another.

melody-rs commented:
@white-axe GitHub won't let me request you as a reviewer (I think I need to add you to the repository collaborators)

white-axe commented on crates/luminol/Cargo.toml:
```diff
@@ -46,12 +46,15 @@ luminol-term.workspace = true
 
 strum.workspace = true
 
+tokio = "1.33" # *sigh*
```
You don't need to rope in a Tokio runtime just to use reqwest. I only added reqwest because I couldn't get surf to compile in web builds without downgrading wasm-bindgen (surf depends on stdweb which is affected by https://github.com/koute/stdweb/issues/427, and eframe requires wasm-bindgen >= 0.2.87) and I don't think the web version of reqwest needs Tokio, so you can switch back to surf in native builds only if you want.

melody-rs replied:
I'd actually like to stick to reqwest because it performs significantly faster than surf it seems

white-axe commented on crates/luminol-graphics/src/sprite/shader.rs:
```diff
@@ -0,0 +1,158 @@
+    bind_group_layouts: &crate::BindGroupLayouts,
+    target: wgpu::BlendState,
+) -> wgpu::RenderPipeline {
+    let use_push_constants = render_state
+        .device
+        .features()
+        .contains(wgpu::Features::PUSH_CONSTANTS);
```
If use_push_constants is going to be calculated from inside of luminol-graphics then ideally all calculations of this variable should be moved to luminol-graphics for consistency. Currently it's also being calculated from luminol-components.

white-axe commented on crates/luminol-core/src/window.rs:
```diff
@@ -0,0 +1,179 @@
+        self.add_boxed_window(Box::new(window))
+    }
+
+    fn add_boxed_window(&mut self, window: Box<dyn Window>) {
+        // FIXME use a hashmap, or something with less than O(n) search time
+        if self.windows.iter().any(|w| w.id() == window.id()) {
+            return;
+        }
+        self.windows.push(window)
+    }
```
Who cares if this is O(n)? There's no way the user is going to add more than one window or tab per frame.

If it really matters that much, you can compute a hashmap of IDs at the beginning of the `display` method, which doesn't lose performance since we need to iterate over the windows once per frame just to draw them.

melody-rs replied:
Good point

white-axe commented:
It seems the egui_dock upgrade introduced some bugs:

* The application crashes when using Close Project while a tab that requires filesystem is undocked into a window
* The script editor window continuously expands vertically when undocking a script editor tab

I'll be fixing these over in #51 so you don't need to worry about them.

white-axe commented on crates/luminol/src/app/top_bar.rs:
```diff
@@ -0,0 +1,322 @@
+                    if let Err(why) = update_state.data.load(
+                        update_state.filesystem,
+                        // TODO code jank
+                        update_state.project_config.as_mut().unwrap(),
```
The application panics here with "called `Option::unwrap()` on a `None` value" when loading a project using the File menu in the top bar. Project loading using the button in the Get Started tab works, however!

melody-rs replied:
I thought I fixed this, but apparently not? I'll look into it again

melody-rs committed:
refactor: :construction: try splitting up the map tab

melody-rs committed:
Merge remote-tracking branch 'upstream/dev' into split-crates

white-axe commented:
I'm planning (after this pull request is done) on refactoring some of the web code I wrote, like by splitting src/web/web_worker_runner.rs and src/filesystem/web.rs into smaller files and using better error handling in the web filesystem. I think it's only fair that I should be responsible for making the web code less bad since I wrote the code two weeks ago and therefore I remember a lot about it. So please don't stress out too much when you get to the WebAssembly support.

melody-rs commented:
:+1:

melody-rs committed:
Enable -Zthreads compiler flag

melody-rs committed:
Update nightly in workflows

melody-rs committed:
revert: :rewind: Undo making cursor state an enum (will tackle this later)

melody-rs committed:
Sorta resolve dependencies on wasm

Still a mess (and is especially complicated by the fact that you can't specify workspace target specific dependencies)

melody-rs committed:
Fix backing web filesystem implementation

Still need to work on the project filesystem though!
I'll be honest, I'm not sure how this will work, especially with the way I have the native filesystem set up

melody-rs committed:
Fix native build

melody-rs committed:
Remove jobs flag

melody-rs committed:
Fix audio on wasm

melody-rs committed:
Temporarily impl Send + Sync for wgpu callbacks

In wgpu 0.17 wgpu types are not Send + Sync, because they reference things in the JS heap (meaning they cannot leave the thread they were created on)

egui_wgpu's CallbackTrait requires Send + Sync even on wasm and we need to store wgpu types in callbacks. The callback_resources typemap passed to callbacks doesn't need to be Send + Sync.. but I can't find a simple way to add anything to the typemap.

https://github.com/emilk/egui/issues/1399 feels relevant

melody-rs committed:
Fix winit not compiling in CI

https://github.com/emilk/egui/pull/3228

melody-rs committed:
Get wasm filesystem compiling

melody-rs committed:
Get luminol compiling on wasm!

melody-rs committed:
Fix missed rename

melody-rs committed:
Dumb typo

melody-rs committed:
Move luminol crate to be root package

Trunk doesn't like virtual workspaces unfortunately
:(

melody-rs committed:
Run workspace tests

white-axe commented on crates/luminol-filesystem/src/web.rs:
```diff
@@ -43,127 +42,119 @@ pub struct FileSystemCommand(FileSystemCommandInner);
+            return Err(Error::Wasm32FilesystemNotSupported);
         }
-        let (oneshot_tx, oneshot_rx) = oneshot::channel();
+        let (oneshot_tx, oneshot_rx) = flume::bounded(1);
```
Might be better to use actual oneshot channels since they're implemented a lot more efficiently. [async-oneshot](https://crates.io/crates/async-oneshot) seems nice. Its receivers don't have a blocking receive method, but you can do that by using `futures::executor::block_on` or by using a different, synchronous oneshot channel implementation for those cases.

melody-rs replied:
I'll try using the `oneshot` crate as it supports both async and sync channels

white-axe commented on crates/luminol-audio/src/wrapper.rs:
```diff
@@ -82,117 +80,102 @@ impl AudioWrapper {
-        let length = state!()
-            .filesystem
-            .metadata(path)
-            .map_err(|e| e.to_string())?
-            .size as usize;
-        let mut vec = vec![0; length];
-        file.read(&mut vec[..]).map_err(|e| e.to_string())?;
+        let file = filesystem.open_file(path, luminol_filesystem::OpenFlags::Read)?;
```
Changing this part of the code like this to pass the file handle directly to Audio running on the main thread is going to cause a crash because the read method on this file handle is blocking. I can't think of a better way around this than reading the entire file beforehand.

melody-rs replied:
Oh, that's annoying. I think I can fix this pretty easily by relaxing `Audio` to only take something implementing `Read` + `Seek` rather than `File` and then just passing a cursor

melody-rs committed:
fix(audio): :bug: Completely read audio file on wasm

melody-rs committed:
perf(wasm): :zap: Use oneshot crate for oneshot channels

white-axe commented on crates/luminol-ui/src/tabs/started.rs:
```diff
@@ -0,0 +1,199 @@
+                    ));
+                }
+
+                #[cfg(target_arch = "wasm32")]
+                {
+                    self.load_filesystem_promise = Some(poll_promise::Promise::spawn_local(
+                        luminol_filesystem::host::FileSystem::from_idb_key(path),
+                    ));
+                }
```
In order for loading from the recent projects menu to work in web, `from_idb_key` needs to be passed `idb_key`, not `path`.

melody-rs committed:
fix(filesystem): :bug: Pass idb key instead of path to Filesystem::from_idb_key

white-axe commented:
Does the map editor work in web builds yet? When I try opening the map editor I start getting WebGPU errors and the screen turns blank.

melody-rs commented:
It did before, I'm not sure why that's happening? I have a sneaking suspicion it has something to do with `use_push_constants` not affecting shaders properly

I was going to pull in naga_oil to replace the const_format stuff in a separate PR but I'll try it now and see what happens

melody-rs committed:
refactor(tilemap): :recycle: Use naga oil instead of const_format

melody-rs commented:
@white-axe Should be fixed now!

melody-rs committed:
fix(ui): :bug: Fix top bar opening & closing projects

melody-rs committed:
Merge branch 'dev' into split-crates

melody-rs committed:
Merge branch 'dev' into split-crates

melody-rs committed:
Bump nightly in trunk build

somedevfox reviewed (changes requested):
In my opinion, `luminol-` prefix should be omitted in the `crates` directory. The fact that they belong to Luminol is self-explanatory because of the root package manifest and git repository these crates are in.
i.e.
```
crates
└ ━> app
└ ━> audio
└ ━> components
└ ━> [...]
```

melody-rs commented:
Ok, I'll do that! I'll still keep the prefix in the crate names though to avoid collision with module names and because I intend to upload luminol to crates.io in the future.

melody-rs committed:
Remove luminol- in folder prefixes
